rest_tools/server/arghandler.py

```diff
@@ -0,0 +1,153 @@
+
+
+NO_DEFAULT = _NoDefaultValue()
+
+
+def _get_json_body(request_handler: tornado.web.RequestHandler) -> Dict[str, Any]:
+    json_body = json_decode(request_handler.request.body)  # type: ignore[no-untyped-call]
+    return cast(Dict[str, Any], json_body)
+
+
+class _UnqualifiedArgumentError(Exception):
+    """Raise when ArgumentHandler._qualify_argument() fails."""
+
+
+def _make_400_error(arg_name: str, error: Exception) -> tornado.web.HTTPError:
+    if isinstance(error, tornado.web.MissingArgumentError):
+        return error  # MissingArgumentError is already a 400 error
+    return tornado.web.HTTPError(400, reason=f"`{arg_name}`: {error}")
+
+
+class ArgumentHandler:
+    """Helper class for argument parsing, defaulting, and casting.
+
+    Like argparse.ArgumentParser, but for REST & JSON-body arguments.
+    """
+
+    @staticmethod
+    def _qualify_argument(
+        type_: Optional[type], choices: Optional[List[Any]], value: Any
+    ) -> Any:
+        """Cast `value` to `type_` type, and/or check `value` in in `choices`.
+
+        Raise _UnqualifiedArgumentError if either qualification fails.
+        """
+        if type_:
+            try:
+                if isinstance(value, str) and (type_ == bool) and (value != ""):
+                    value = bool(distutils.util.strtobool(value))
+                else:
+                    value = type_(value)
+            except ValueError as e:
+                raise _UnqualifiedArgumentError(f"(ValueError) {e}")
+
+        if choices and (value not in choices):
+            raise _UnqualifiedArgumentError(
+                f"(ValueError) {value} not in options ({choices})"
+            )
+
+        return value
+
+    @staticmethod
+    def _type_check(type_: Optional[type], value: Any) -> None:
+        # check the value's type (None is okay too)
+        if not type_:
+            return
+        if not isinstance(value, type_) and (value is not None):
+            raise ValueError(
+                f"Value, `{value}` ({type(value)}), is not {type_} or None"
+            )
+
+    @staticmethod
+    def get_json_body_argument(  # pylint: disable=R0913
+        request_handler: tornado.web.RequestHandler,
+        name: str,
+        default: Any,
+        choices: Optional[List[Any]],
+    ) -> Any:
+        """Return the argument from JSON-decoded request body."""
+        try:
+            value = _get_json_body(request_handler)[name]
+            return ArgumentHandler._qualify_argument(None, choices, value)
+        except (KeyError, json.decoder.JSONDecodeError):
+            # Required -> raise 400
+            if isinstance(default, type(NO_DEFAULT)):
+                raise _make_400_error(name, tornado.web.MissingArgumentError(name))
+        except _UnqualifiedArgumentError as e:
+            raise _make_400_error(name, e)
+
+        # Else:
+        # Optional / Default
+        try:
+            return ArgumentHandler._qualify_argument(None, choices, default)
```

Comment:
Woe to the poor user who specified a default that wasn't one of their enumerated choices. :laughing:

Reply:
same goes for `ArgumentHandler._type_check(type_, default)`. I've really covered all this bases here